pkg/controller/jobframework/reconciler.go

```diff
@@ -572,66 +569,80 @@ func (r *JobReconciler) recordAdmissionCheckUpdate(wl *kueue.Workload, job Gener
+	if QueueNameForObject(ancestorJob) == "" {
+		return nil, nil
 	}
 	wlList := kueue.WorkloadList{}
-	if err := r.client.List(ctx, &wlList, client.InNamespace(ancestor.GetNamespace()), client.MatchingFields{indexer.OwnerReferenceUID: string(ancestor.GetUID())}); client.IgnoreNotFound(err) != nil {
+	if err := r.client.List(ctx, &wlList, client.InNamespace(ancestorJob.GetNamespace()),
+		client.MatchingFields{indexer.OwnerReferenceUID: string(ancestorJob.GetUID())},
+	); client.IgnoreNotFound(err) != nil {
 		return nil, err
 	}
 	if len(wlList.Items) > 0 {
 		// In theory the job can own multiple Workloads, we cannot do too much about it, maybe log it.
+		ctrl.LoggerFrom(ctx).V(2).Info(
+			"WARNING: The job has multiple associated Workloads.",
+			"job", ancestorJob.GetName(),
+			"workloads", klog.KObjSlice(wlList.Items),
+		)
 		return &wlList.Items[0], nil
 	}
 	return nil, nil
 }
 
-// getAncestorJobManagedByKueue traverses controllerRefs to find an ancestor job that is manged by Kueue (ie, it has a queue-name label).
-func (r *JobReconciler) getAncestorJobManagedByKueue(ctx context.Context, jobObj client.Object) (client.Object, error) {
-	seen := sets.New[types.UID]()
-	currentJob := jobObj
+func GetTopLevelAncestorJobManagedByKueue(ctx context.Context, c client.Client, record record.EventRecorder, jobObj client.Object) (client.Object, error) {
+	var (
+		log         = ctrl.LoggerFrom(ctx)
+		seen        = sets.New[types.UID]()
+		currentObj  = jobObj
+		topLevelJob client.Object
+	)
+
 	for {
-		if seen.Has(currentJob.GetUID()) {
-			return nil, nil
+		if seen.Has(currentObj.GetUID()) {
+			log.Error(ErrCyclicOwnership,
+				"Terminated search for Kueue-managed Job because of cyclic ownership",
+				"owner", currentObj,
+			)
+			return nil, ErrCyclicOwnership
 		}
-		seen.Insert(currentJob.GetUID())
+		seen.Insert(currentObj.GetUID())
 
-		owner := metav1.GetControllerOf(currentJob)
-		if owner == nil || !IsOwnerManagedByKueue(owner) {
-			return nil, nil
+		owner := metav1.GetControllerOf(currentObj)
+		if owner == nil {
+			return topLevelJob, nil
 		}
-		parentJob := GetEmptyOwnerObject(owner)
-		if parentJob == nil {
-			return nil, fmt.Errorf("workload owner %v: %w", owner, ErrUnknownWorkloadOwner)
+
+		parentObj := GetEmptyOwnerObject(owner)
+		managed := parentObj != nil
+		if parentObj == nil {
+			parentObj = &metav1.PartialObjectMetadata{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: owner.APIVersion,
+					Kind:       owner.Kind,
+				},
+			}
 		}
-		if err := r.client.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: jobObj.GetNamespace()}, parentJob); err != nil {
+		if err := c.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: jobObj.GetNamespace()}, parentObj); err != nil {
 			return nil, errors.Join(ErrWorkloadOwnerNotFound, err)
 		}
-		if QueueNameForObject(parentJob) != "" {
-			return parentJob, nil
+		// Ignore unmanaged resources and disabled jobs.
```

Comment:
I don't understand the comment, and it does not match the checks, unless I'm missing something. It says "Ignore unmanaged", but checks "managed". Also, I don't understand what is "resources" in this context.

Reply:
I think this was missing in the previous logic. I've removed the comment.